pdns/dnsdist-lua-bindings-dnsquestion.cc

```diff
@@ -54,6 +54,27 @@ void setupLuaBindingsDNSQuestion()
   g_lua.registerFunction<bool(DNSQuestion::*)()>("getDO", [](const DNSQuestion& dq) {
       return getEDNSZ((const char*)dq.dh, dq.len) & EDNS_HEADER_FLAG_DO;
     });
+  g_lua.registerFunction<vector<uint8_t>(DNSQuestion::*)(void)>("getTrailingData", [](const DNSQuestion& dq) {
+      const uint8_t* message = reinterpret_cast<const uint8_t*>(dq.dh);
+      const uint16_t length = getDNSPacketLength(reinterpret_cast<const char*>(message), dq.len);
+      vector<uint8_t> tail(message + length, message + dq.len);
+      return tail;
+    });
+  g_lua.registerFunction<bool(DNSQuestion::*)(vector<pair<int, uint8_t>>)>("setTrailingData", [](DNSQuestion& dq, const vector<pair<int, uint8_t>>&data) {
+      uint8_t* message = reinterpret_cast<uint8_t*>(dq.dh);
+      const uint16_t length = getDNSPacketLength(reinterpret_cast<const char*>(message), dq.len);
+      if(length + data.size() > dq.size) {
```

Comment:
In TCP, it looks like the buffer size is only a little bit bigger than the query size, and exactly the same size as the response size:
```c++
        /* allocate a bit more memory to be able to spoof the content,
           or to add ECS without allocating a new buffer */
        queryBuffer.resize(qlen + 512);
```
And:
```c++
        size_t responseSize = rlen;
        uint16_t addRoom = 0;
#ifdef HAVE_DNSCRYPT
        if (dnsCryptQuery && (UINT16_MAX - rlen) > (uint16_t) DNSCRYPT_MAX_RESPONSE_PADDING_AND_MAC_SIZE) {
          addRoom = DNSCRYPT_MAX_RESPONSE_PADDING_AND_MAC_SIZE;
        }
#endif
        responseSize += addRoom;
        answerBuffer.resize(responseSize);
```
Both in `dnsdist-tcp.cc`. 

In UDP, this seems to be a fixed 4096-bytes, derived from MMReciever:
```c++
  struct MMReceiver
  {
    char packet[4096];
    /* used by HarvestDestinationAddress */
    char cbuf[256];
    ComboAddress remote;
    ComboAddress dest;
    struct iovec iov;
  };
        .
        .
        .
      processUDPQuery(*cs, holders, msgh, remote, recvData[msgIdx].dest, recvData[msgIdx].packet, static_cast<uint16_t>(got), sizeof(recvData[msgIdx].packet), outMsgVec.get(), &msgsToSend, &recvData[msgIdx].iov, recvData[msgIdx].cbuf);
```

Personally I think it is okay to resize the buffer on these cases, but to do this safely we'd have to:

1. Track the fact that we have allocated a new buffer for this instead of using the one passed on construction
2. Add a destructor that frees up the memory if allocated

It is a much more invasive change (not because of the amount of code, but because adding a destructor to this means looking through the code to see everywhere that it is used and making sure that it is used properly), and also one that would probably raise eyebrows on the PowerDNS side.

I have no great advice here. It seems a sucky situation.

Reply:
It seems to me like we can't enlarge the buffer inside this function, because we don't have access to pointers referencing it (cf. [`fixUpResponse`](https://github.com/gibson042/pdns/blob/d4de01b29046ca79ca2edd000187f71e5effc005/pdns/dnsdist.cc#L296-L303)). For TCP, I a symmetry argument may convince them to increase the default response `addRoom` from 0 to 512 (as it is with queries). For UDP, 4096 is just their self-imposed limit (almost certainly from the [recommendation in RFC 6891](https://tools.ietf.org/html/rfc6891#section-6.2.5)) and I think it's fine.

Fortunately, none of these limits will be a problem for us because we're only mutating queries, not supporting large ones, and adding less than 512 octets.